blog: add a post on python source analysis goal (#136)
pantsbuild/pantsbuild.org: blog/2024-01-25-inspecting-dependency-inference-results/index.mdx

---
    authors: [alexey]
    tags: [concepts]
---

# Inspecting dependency inference results

<CaptionedImg src={require("./splash.jpg").default}>
  Photo by [Agence
  Olloweb](https://unsplash.com/photos/magnifying-glass-near-gray-laptop-computer-d9ILr-dbEdg)
  /
  [Unsplash](https://unsplash.com/photos/magnifying-glass-near-gray-laptop-computer-d9ILr-dbEdg)
</CaptionedImg>

{/* truncate */}

> An important distinguishing feature of Pants v2 is its ability to automatically [infer your code's internal and external dependencies](https://www.pantsbuild.org/blog/2022/10/27/why-dependency-inference). The information about them is available via the `dependencies` and `peek` goals, so that metadata about what modules import from modules is always readily accessible.

Reviewer: ```suggestion
An important distinguishing feature of Pants v2 is its ability to automatically [infer your code's internal and external dependencies](https://www.pantsbuild.org/blog/2022/10/27/why-dependency-inference). The information about those inferences is available via the `dependencies` and `peek` goals, so that metadata about what modules import from other modules is always readily accessible to you.
```

> While managing a codebase in a language supported by Pants such as Python, Java, or Scala, it can, however, be useful to go further and explore your import statements to compare them with what has been inferred by Pants. This may be particularly helpful for large modules with lots of imports where it's easy to miss the fact that a dependency is omitted or that there's a dependency inferred that should really be ignored. To get help with that, you can use one of the less known goals, [`python-dump-source-analysis`](https://www.pantsbuild.org/2.18/reference/goals/python-dump-source-analysis), which can dump source code analysis for the Python source code build targets.

Reviewer: ```suggestion
While managing a codebase in a [Pants-supported language](https://www.pantsbuild.org/2.18/docs/introduction/welcome-to-pants#which-languages-and-frameworks-does-pants-support) (e.g. Python, Java, Scala, etc...), it can sometimes be useful dig deeper and explore the inferred dependencies in order to compare them to your source code's import statements. This can be particularly useful for large modules with many imports, where it's very easy to miss an omitted dependency or an inferred dependency that should have been ignored. 

You can start your exploration with one of the lesser known goals, [`python-dump-source-analysis`](https://www.pantsbuild.org/2.18/reference/goals/python-dump-source-analysis). This goal will dump the dependency analysis for the Python source code build targets.
```

> This command may be used as a compliment to any static code analysis tooling you may already employ such as [semgrep](https://semgrep.dev/) or [flake8](https://flake8.pycqa.org/en/latest/) as they can only parse your modules semantically and often won't be able to link individual imports to build targets in your codebase.

Reviewer: ```suggestion
This command may be used to compliment any static code analysis tooling you may already employ such as [semgrep](https://semgrep.dev/) or [flake8](https://flake8.pycqa.org/en/latest/), as they can only parse your modules semantically and often won't be able to link individual imports to build targets in your codebase.
```

> In order to run this goal, you need to enable the `"pants.backend.experimental.python"` backend:

Reviewer: Why is it in experimental?

Author: Historical reasons, I guess - no one paid much attention to it...

Reviewer 2: It hasn't been promoted to non-experimental. I think we'd want to standardise with the other goals at least for the default analysis flavour

```
[
  {
  ...
    "identified": {
      "imports": {
        "json": {
          "lineno": 1,
          "weak": false
        },
        "typing.Optional": {
          "lineno": 3,
          "weak": false
        },
        "requests": {
          "lineno": 5,
          "weak": false
        },
        "cheeseshop.repository.package.Package": {
          "lineno": 9,
          "weak": false
        }
      }
  ...
```

You can also take a look at the resolve status for every `import` statement. For instance, unownable imports are those that cannot be owned by a build target; these are usually standard library imports. Unowned imports are those that Pants could not resolve; this could be due to an attempt to import a non-existing first party module or a third party package that was not listed in the project requirements and Pants isn't aware of it.

> Keep in mind that Pants would only infer dependencies when it is safe to as it checks for ambiguity. For instance, if you have multiple package requirements declared in the same Python resolve, e.g. `requests==2.25.1` and `requests==2.22.0`, Pants won't infer a dependency and will instead provide a suggestion for how to disambiguate. Unambiguous imports are those that are owned by exactly one build target whereas ambiguous ones are those that cannot be linked to a specific build target. See a partial output of the command below:

Reviewer: ```suggestion
Keep in mind that Pants would only infer dependencies when it is safe to - as it checks for ambiguity. For instance, if you have multiple package requirements declared in the same Python resolve, e.g. `requests==2.25.1` and `requests==2.22.0`, Pants won't infer a dependency and will instead provide a suggestion for how to disambiguate. Unambiguous imports are those that are owned by exactly one build target whereas ambiguous ones are those that cannot be linked to a specific build target. See a partial output of the command below:
```

          "address": [
            "requirements#requests"
          ]
        },
        "cheeseshop.repository.package.TypoInModule": {
          "status": "ImportOwnerStatus.unowned",
          "address": []
        },
        "cheeseshop.repository.parsing.exceptions.PackageNotFoundError": {
          "status": "ImportOwnerStatus.unambiguous",
          "address": [
            "cheeseshop/repository/parsing/exceptions.py"
          ]
        }
      },
    ...
```

Moreover, Pants can infer a target's dependencies based on strings that look like dynamic dependencies, such as Django settings files expressing dependencies as strings or `pytest` plugins listed in the `pytest_plugins` variable in a test module or a `conftest.py` file. These dependencies are considered to be "weak" and are reported separately:

```shell-session
$ pants python-dump-source-analysis \
  --analysis-flavor=raw_dependency_inference \
  cheeseshop/repository/repository.py > sources.json

$ jq -r '.[].identified.imports' sources.json \
    | jq 'map_values(select(.weak == true))'
{
  "cheeseshop.repository.properties": {
    "lineno": 12,
    "weak": true
  }
}
```

> With this goal, you can also check if a guarded import got ignored. For instance, the import in

Reviewer: ```suggestion
With this goal, you can also check if a guarded import was ignored. For instance, the import in:
```
